contributors/design-proposals/sig-cli/kubectl_apply_getsetdiff_last_applied_config.md

```diff
@@ -0,0 +1,181 @@
+an object, it must be manually removed from the annotation and then be removed
+from the local object configuration.  Users should be able to simply edit
+the local object configuration and set it as the last-applied-config
+to be used for the next diff base.  Storing the last-applied-config
+in an annotation adds black magic to `kubectl apply`, and it would
+help users learn and understand if the value was exposed in a discoverable
+manner.
+
+## Use Cases
+
+1. As a user, I want to be able to diff the last-applied-configuration
+against the current local configuration to see which changes the command is seeing
+2. As a user, I want to remove fields from being managed by the local
+object configuration by removing them from the local object configuration
+and setting the last-applied-configuration to match.
+3. As a user, I want to be able to view the the base
+last-applied-configuration that will be used to merge changes
+
+## Naming and Format possibilities
+
+### Naming
+
+1. last-applied
+2. last-config
+3. last-applied-config
+4. last-configuration
+5. last-applied-configuration
+6. last
+
+### Formats
+
+1. Apply subcommands
+  - `kubectl apply set/view/diff last-applied
+  - a little bit odd to have 2 verbs in a row
+  - improves discoverability to have these as subcommands so they are tied to apply
+2. Set/View subcommands
+  - `kubectl set/view/diff last-applied
+  - consistent with other set/view commands
+  - clutters discoverability of set/view commands since these are only for apply
+  - clutters discoverability for last-applied commands since they are for apply
+3. Apply flags
+  - `kubectl apply [--set-last-applied | --view-last-applied | --diff-last-applied]
+  - Not a fan of these
+
+## view last-applied
+
+Porcelain command that retrieves the object and prints the annotation value as yaml or json.
+
+Prints an error message if the object is not managed by `apply`.
+
+1. Get the last-applied by type/name
+
+```sh
+kubectl apply view last-applied deployment/nginx
+```
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: nginx
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        run: nginx
+    spec:
+      containers:
+      - image: nginx
+        name: nginx
+```
+
+2. Get the last-applied by file, print as json
+
+```sh
+kubectl apply view last-applied -f deployment_nginx.yaml -o json
+```
+
+Same as above, but in json
+
+## diff last-applied
+
+Porcelain command that retrieves the object and displays a diff against the local configuration
+
+1. Diff the last-applied
+
+```sh
+kubectl apply diff last-applied -f deployment_nginx.yaml
+```
+
+Opens up a 2-way diff in the default diff viewer.  last-applied is old, local configuration is new.
+This is meant to show the user what they changed in the configuration,
+since it was last applied, but not show what has changed in the server.
+
+A future goal is to provide a 3-way diff with `kubectl apply diff -f deployment_nginx.yaml`.
```

Comment:
IIUC, 3-way diff should be the final patch we will send to server. am I correct?